### PR TITLE
MINOR: Update to Gradle 6.3

### PR DESCRIPTION
**Pull request by ijuma**
* Introduce `gradlewAll` script to replace `*All` tasks since the approach
used by the latter doesn't work since Gradle 6.0 and it's unclear when,
if ever, it will work again ( see https://github.com/gradle/gradle/issues/11301).
* Update release script and README given the above.
* Update zinc to 1.3.5.
* Update gradle-versions-plugin to 0.28.0.

The major improvements in Gradle 6.0 to 6.3 are:
- Improved incremental compilation for Scala
- Support for Java 14 (although some Gradle plugins
like spotBugs may need to be updated or disabled,
will do that separately)
- Improved scalac reporting, warnings are clearly
marked as such, which is very helpful.

Tested `gradlewAll` manually for the commands listed in the README
and release script. For `uploadArchive`, I tested it with a local Maven
repository.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ijuma** [comment]
Currently blocked by https://issues.apache.org/jira/browse/INFRA-19418

**Comment by ijuma** [comment]
retest this please

**Comment by ijuma** [comment]
retest this please

**Comment by ijuma** [comment]
retest this please

**Comment by ijuma** [comment]
retest this please

**Comment by omkreddy** [comment]
@ijuma  I am getting errors when I try  "all" targets.  not sure if this is due to gradle 6 or my env issue.

```
 ./gradlew clean jarAll

> Configure project :
Building project 'core' with Scala version 2.12.10
Building project 'streams-scala' with Scala version 2.12.10

> Task :jarScala_2_12 FAILED

Execution failed for task ':jarScala_2_12'.
> Included build /Users/mkumar/workspace/kafka has build path :kafka which is the same as included build /Users/mkumar/workspace/kafka
```

**Comment by ijuma** [comment]
@omkreddy Thanks for testing. It looks like that issue happens with trunk too. Is that the case for you too?

**Comment by omkreddy** [comment]
@ijuma It happened once.  But after killing Gradle Daemon process, it never happened on trunk (ran multiple times). 

some related links:
https://github.com/gradle/gradle/issues/11301
https://docs.gradle.org/current/userguide/upgrading_version_5.html#changes_to_build_and_task_names_in_composite_builds

**Comment by ijuma** [comment]
This is blocked due to the Gradle issue @omkreddy mentioned. I included some of the changes in this PR in #7909. Once that's merged, I will rebase this branch and it should be only the Gradle version bump.

**Comment by ijuma** [comment]
@omkreddy Given that the Gradle issue doesn't seem to be resolved anytime soon, I moved the logic to run a gradle command with multiple Scala versions to a `gradlewAll` script. What do you think?

File: README.md
Comment (omkreddy) [inline README.md:94]: We also need to update `./gradlew uploadArchivesAll`, `./gradlew installAll` references below

File: release.py
Comment (omkreddy) [inline release.py:601]: need to update `./gradlew uploadArchivesAll` at line no: 644

**Review by omkreddy — APPROVED** [review]
@ijuma Thanks for the update. LGTM.  I assume we have tested the changes locally.

**Comment by chia7712** [comment]
The scala compilation get slower after applying this PR ( 3m50.56s -> 4m49.70s). The build command is ```./gradlew clean build -x test```. Not sure whether this happens on my local only.

related issue: https://github.com/gradle/gradle/issues/12591

**Comment by ijuma** [comment]
@chia7712 That's because you have `clean` in the command. The benefit of using `zinc` is that incremental recompilation is much faster. If you try the same test where you just change one or two files and don't run `clean`, you should see a significant perf improvement.

**Comment by ijuma** [comment]
One job passed and the other failed with a flaky test failure:

> org.apache.kafka.connect.mirror.MirrorConnectorsIntegrationTest.testReplication

**Comment by chia7712** [comment]
> That's because you have clean in the command. The benefit of using zinc is that incremental recompilation is much faster. If you try the same test where you just change one or two files and don't run clean, you should see a significant perf improvement.

thanks for the explanation. My comment was not clear. The question was the clean + build get slower after updating gradle to 6.x (unrelated to this PR. the only change is gradle version). According to the gradle profile report, the duration of compiling scala module become higher ( core: 1m38.90s -> 2m11.35s, streams:streams-scala: 4.525s	-> 23.316s). By contrast, the duration of compiling java modules are almost same. loop 5 times and it seems the duration is consistently higher on gradle 6.x

**Comment by ijuma** [comment]
@chia7712 I understood. :) Incremental compilation tracking has a cost. The new version of Gradle uses a new version of Zinc which trades full compilation speed for incremental compilation speed. I think that's fine given that incremental compilation is what developers care most about.

**Comment by ijuma** [comment]
> I assume we have tested the changes locally.

I wanted to get some feedback on the approach, so I had just sanity checked it. I will do more thorough verification before merging.

**Comment by lure** [comment]
@ijuma, would you be so kind to provide a link on that zinc changes if you have it somewhere near you? 
Thank you.

Upd: got an answer from  Guillaume Martres on https://gitter.im/sbt/zinc-contrib they say it is not how zinc behaves, and most probably is bug in Gradle.

**Comment by ijuma** [comment]
@lure In the past, sbt with incremental compiler vs ant has shown similar behavior where the former was faster under incremental compilation while ant was faster under clean builds. The comment in that chat says the build time doubled, but that's not what was reported here (`3m50.56s -> 4m49.70s`).

**Comment by lure** [comment]
That's right, my case shows nearly x2, i.e. 6mins become 11. 
1 minute is not expected 2% too. I mean, there shouldn't be slowing at all, new Zinc promises speed up instead. 

In my case the problem is reproducible just by replacing Gradle version from 5.6.2 to 6.3.  

@chia7712 located particular change that introduced regress, I believe?

Will check my project module by module, may be it is something else.

**Comment by ijuma** [comment]
@lure This is the issue I was thinking of where things regressed previously with better incremental recompilation:

https://github.com/gradle/gradle/issues/1475#issuecomment-283932472

It does look like this time, it may be a Gradle regression in 6.0 as per the issue @chia7712 linked to.

**Comment by lure** [comment]
It would be great if this is just a regress. :) The issue I observe is reproducible on 6.2 too so it could be the case. 
We do not use ant/parallel options and it's 5 to 6 only for us. 
Thank you for explanation.

**Comment by ijuma** [comment]
I tested `./gradlew install` with trunk versus this PR after `./gradlew clean`:

* trunk: BUILD SUCCESSFUL in 1m 23s
* this PR: BUILD SUCCESSFUL in 1m 50s

I think that's OK given the incremental compilation improvements.

**Comment by ijuma** [comment]
@omkreddy I've done a bunch of testing and this seems to work as expected generally.

One part that I wasn't completely sure about is the behavior of `./gradlew uploadArchivesAll` in trunk versus `./gradlewAll uploadArchives` in this PR with regards to Java modules that end up getting built twice due to being dependencies of the Scala modules (e.g kafka-clients, kafka-streams). When I try this in trunk with a local Maven repository, it fails. With the code in this PR, it completes, but these modules get uploaded twice.

So, I have decided to go ahead and merge this PR.

**Pull request merged**
